### PR TITLE
Restore previous behaviour of poUsePath (by adding posixFindExe)

### PR DESCRIPTION
**Pull request by zielmicha**
Related to #6299

### CODE REVIEW

**Comment by zielmicha** [comment]
Any comments? 

I think that it's bad enough that 0.17.2 was released with broken `osproc`.

**Comment by dom96** [comment]
> I think that it's bad enough that 0.17.2 was released with broken osproc.

In what way?

I'm not at all up to date about this. The fact that we need two different ``findExe``'s sucks, what is the use case for this?

**Comment by zielmicha** [comment]
Currently poUsePath uses current directory + PATH to find executable, even on Linux systems. This is, at least, a breaking change.

**Comment by Araq** [comment]
There is no breaking change. The code used to work this way before. How do I know? Because I tried to not look at the current dir and things broke.

**Comment by dom96** [comment]
Okay, I think I understand. Comments to follow.

File: lib/pure/osproc.nim [hunk 1]
```diff
@@ -747,6 +747,29 @@ elif not defined(useNimRtl):
     tags: [ExecIOEffect, ReadEnvEffect, ReadDirEffect, RootEffect], cdecl, gcsafe.}
   {.pop.}
 
+  proc posixFindExe*(exe: string): string {.
```
Comment (dom96) [inline lib/pure/osproc.nim]: Do we really need another proc for this? Wouldn't it be easier to add an option to ``findExe`` for this behaviour?

File: lib/pure/osproc.nim [hunk 2]
```diff
@@ -952,18 +978,7 @@ elif not defined(useNimRtl):
     discard close(data.pErrorPipe[readIdx])
     discard fcntl(data.pErrorPipe[writeIdx], F_SETFD, FD_CLOEXEC)
 
-    if data.optionPoUsePath:
-      when defined(uClibc) or defined(linux):
-        # uClibc environment (OpenWrt included) doesn't have the full execvpe
-        let exe = findExe(data.sysCommand)
-        discard execve(exe, data.sysArgs, data.sysEnv)
-      else:
```
Comment (dom96) [inline lib/pure/osproc.nim]: I'm not sure what this code does but seeing you simplify it like this makes me uncomfortable, are you sure you're not introducing any bugs here?

File: tests/osproc/tpath.nim [hunk 3]
```diff
@@ -0,0 +1,44 @@
+discard """
+  output: ""
```
Comment (dom96) [inline tests/osproc/tpath.nim]: Nit: you can be more direct here and specify ``action: run``, for more info see here: https://github.com/nim-lang/Nim/tree/devel/tests#action

File: tests/osproc/tpath.nim [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    discard posixFindExe("exe")
+
+  putEnv("PATH", "/tmp/nim-osproc-test/1")
+  expect(OSError):
+    discard posixFindExe("exe")
+
+  putEnv("PATH", "/tmp/nim-osproc-test/1:/tmp/nim-osproc-test/3")
+  expect(OSError):
+    discard posixFindExe("exe")
+
```
Comment (dom96) [inline tests/osproc/tpath.nim]: While you're at it, maybe you could test osproc's ``poUsePath`` option here as well?

Comment (dom96) [reply tests/osproc/tpath.nim]: Also, are you testing whether ``posixFindExe`` will find the exe in the current directory anywhere? You should test this.

File: lib/pure/osproc.nim (hunk 2 above)
Comment (zielmicha) [reply lib/pure/osproc.nim]: I believe not - this was workaround for uClibc not having execvpe - now osproc doesn't use it at all (it implements the `p` functionality with posixFindPath).

File: tests/osproc/tpath.nim (hunk 4 above)
Comment (zielmicha) [reply tests/osproc/tpath.nim]: Tests for both added.

**Comment by zielmicha** [comment]
> it's full of Unix'sms.
I believe that's not a problem - the code it's meant to test Unix specific code.

File: lib/pure/osproc.nim (hunk 1 above)
Comment (dom96) [reply lib/pure/osproc.nim]: I would still prefer an extending findExe instead of creating a separate proc.

Comment (zielmicha) [reply lib/pure/osproc.nim]: It shares almost no code with findExe, so I would just need to add another big `if`, which seems strange.

**Comment by zielmicha** [comment]
Any chances to get this merged? (I can fix the conflict, but I think I have fixed/replied to all issues)

If not, I'll at least submit PR for #6939 ("." added to PATH by default).

File: lib/pure/osproc.nim (hunk 2 above)
Comment (dom96) [reply lib/pure/osproc.nim]: macOS is still using ``execvp`` though. But I see you have made that change 4 years ago.

Just to be safe, I looked into the blame for these lines and the latest modifications were made by @Araq and @def-, can you guys review this?

File: lib/pure/osproc.nim (hunk 1 above)
Comment (zielmicha) [reply lib/pure/osproc.nim]: `posixFindExe` also checks if the file is executable and raises EPERM when it's not. If you don't like this additional check (I've added it for compatibility with all other tools/languages that use execvp), I can merge both functions. Or add I can add `posixMode` flag to `findExe`.

**Comment by Araq** [comment]
I fixed this differently.

**Pull request closed without merging**